### PR TITLE
Synchronize PipelineRun for multi branch pipeline

### LINKED ISSUE #257: Error handling PipelineRun synchronization for multi-branch pipeline
### What happend?

I can not see any pipeline activities triggered by `Branch indexing` in GitLab pipeline.

And I found that the Pipeline synchronizer did a wrong setting on reference name, please see:

```yaml
apiVersion: devops.kubesphere.io/v1alpha3
kind: PipelineRun
metadata:
  annotations:
    devops.kubesphere.io/jenkins-pipelinerun-id: '1'
  generateName: gitlab-pipeline-
  name: gitlab-pipeline-2bf5m
  namespace: my-devops-projectsg945
spec:
  pipelineRef:
    kind: Pipeline
    name: gitlab-pipeline
    namespace: my-devops-projectsg945
  pipelineSpec:
    multi_branch_pipeline:
      discarder:
        days_to_keep: '7'
        num_to_keep: '5'
      gitlab_source:
        credential_id: gitalb-id
        discover_branches: 1
        discover_pr_from_forks:
          strategy: 2
          trust: 2
        discover_pr_from_origin: 2
        discover_tags: true
        owner: johnniang
        repo: johnniang/jenkinsfile-demo
        server_name: 'https://gitlab.com'
      name: gitlab-pipeline
      script_path: Jenkinsfile
      source_type: gitlab
    type: multi-branch-pipeline
  scm:
    refName: 'https://gitlab.com/johnniang/jenkinsfile-demo/-/merge_requests/3'
    refType: ''
```

### Steps to reproduce

1. Install KubeSphere nightly: `nightly-20210923`
1. Create a GitLab pipeline
1. See pipeline activities

/kind bug
/milestone v3.2

### LINKED ISSUE #260: Error  activity synchronization for Github pipeline
`Versions used`
KubeSphere: v3.2.0-alpha.0

To Reproduce
1、Create a GitHub pipeline
2、See pipeline activities

image
image

/kind bug
/cc @kubesphere/sig-devops
/priority high

### PR DESCRIPTION
**Pull request by JohnNiang**
### What this PR dose / why we need it

- Add PipelineRun identity and finder to filter PipelineRun
- Refine PipelineRun creation function

### Which issue(s) this PR fixes:

- Fix #257
- Fix #260

### Steps to test

Docker image for test: `johnniang/devops-controller:sync-pr-with-multi-branch`

### CODE REVIEW

**Comment by JohnNiang** [comment]
hi @LinuxSuRen , how about this way to distinguish two different PipelineRuns:

https://github.com/kubesphere/ks-devops/blob/af4e3d6343d8c020ea70be02dba513de3aa32832/controllers/jenkins/pipelinerun/pipelinerun_synchronizer.go#L41-L67

File: controllers/jenkins/pipelinerun/pipelinerun_finder_test.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+		name: "Multi-branch PipelineRuns",
+		args: args{
+			pipelineRuns: []v1alpha3.PipelineRun{*multiBranchPipeline1, *multiBranchPipeline2},
+		},
+		want: pipelineRunFinder{
+			pipelineRunIdentity{id: "123", refName: "main1"}: multiBranchPipeline1,
+			pipelineRunIdentity{id: "456", refName: "main2"}: multiBranchPipeline2,
+		},
+	}, {
+		name: "Dumplicated multi-branch PipelineRuns",
```
Comment (LinuxSuRen) [inline controllers/jenkins/pipelinerun/pipelinerun_finder_test.go:189]: typo `Dumplicated`

**Comment by LinuxSuRen** [comment]
image

The paging calculate is incorrect.

**Pull request merged**